Finite Build Coils

ryanwu4 opened the pull request.

dpanici commented:
See `CoilCoilMinDistance` objective for masking logic for avoiding doing biot savart of a coil over itself.

ryanwu4 commented:
I identified the 2 issues preventing JAX grad:
1. `safenorm` is necessary if eval points and coil points are the same, which is the case for the centerline self field. I changed the biot savart integral to do that, and also implemented an improved singularity-subtracted integral from the paper which should require fewer quad points, specifically for the compute_self_field function.
2. NaNs appearing in the unregularized biot savart integral were screwing with the gradient as well. I now have `compute_self_field` output a mask of where centerline points are, and the maxNormB objective function ignores those points (in a kinda gross way). I decided to do this over having a biot-savart safemode where it regularizes for points close to a centerline since the latter is more computationally expensive and might not play well with JAX control flow logic.

ryanwu4 commented on desc/coils.py:
```diff
@@ -936,6 +1116,595 @@ def _check_type(coil0, coil):
+
+        names = ["p_frame", "q_frame", "curv1_frame", "curv2_frame", "x"]
+
+        if params is not None:
```
Is the way that I'm packing in params and transforms here correct? This is kind of hacky because the terms `p_frame`, `curv1_frame`, etc should (and are?) only being calculated with the curve implementation, which means that they don't vary over the cross-section dimensions of the grid and only over the centerline dimension. This is the intended behavior, as these variables are only dependent upon the centerline filament, even if the finite build components are dependent on the full cross section geometry.

ryanwu4 replied:
Getting the params with `get_params` and computing over the full grid with `compute_fun` instead of `self.compute()` basically allows me to pass in the full 3D coil grid without overwriting the base behavior of `compute` as implemented in the `Curve` class.

f0uriest replied:
in theory this is fine, but shouldn't params already have the current and centerline dims?

ryanwu4 commented on desc/compute/_field.py:
```diff
@@ -3516,3 +3516,362 @@ def _L_grad_B(params, transforms, profiles, data, **kwargs):
+    a = params["cross_section_dims"][0]
+    b = params["cross_section_dims"][1]
+
+    # kind of a trick, the p and q frame computations are only done on the centerline
```
This is what I was referring to about the centerline computations being done in a weird way, since technically `p_frame` only varies over the centerline dimension, while `u_fb` only varies over the cross section dimension, but the parameterization correctly preserves this convention even with a 3D grid.

ryanwu4 commented:
I'm adding final unit tests on the associated objective functions shortly, but otherwise this PR is basically ready to review.

f0uriest commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+        def biot_savart_subfilament(x):
+            dx = (
+                jnp.roll(x, -1, axis=0) - x
+            )  # using finite diff instead of analytic tangent
```
(the main reason for this is that using approximate tangents will ruin the exponential convergence of the integral making it really innaccurate. If you don't have exact analytic tangents you should use `biot_savart_hh` instead)

ryanwu4 replied:
I believe that the tangent is not exactly the same as the centerline, as in the case of a straight segment of coil with the cross section profile rotating, the tangent vectors of current chunks at the outer edges of the bundle will have a rotational component which the centerline does not have. I'll use `biot_savart_hh `instead.

f0uriest replied:
If it's possible to get the analytic tangents that would be the best option, `biot_savart_quad` is generally a lot more accurate than `biot_savart_hh` for a given resolution, provided the tangents are correct. 

It sounds like the correct tangent would be the centerline tangent + rotation which should be possible.

f0uriest commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+            B = xyz2rpz_vec(B, phi=phi)
+        return B
+
+    def compute_magnetic_field_multifilament_circ(
```
`_private`

f0uriest commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+            Refer to `prep_grid` method. If `prep_grid` is False, this is assumed to
+            have the same toroidal dimension as the xsection grid.
+        coil_frame : bool, optional
+            Whether to compute the field in the t, p ,q coil frame. Default is False,
```
not sure what this means?

f0uriest commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+                finite_build_grid, centerline_grid, coil_frame, params, transforms
+            )
+
+    def compute_self_field_rect(
```
`_private`

f0uriest commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+
+        return B_self, x_fb, centerline_mask
+
+    def compute_self_field_circ(
```
`_private`

f0uriest commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+            "Circular cross section self field not implemented yet"
+        )
+
+    def project_coil_frame(
```
Is this method needed? I feel like comptuting projections is pretty straightforward?

f0uriest commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+        return self._cross_section_shape
+
+    @classmethod
+    def prep_grid(cls, xsection_grid, centerline_grid=None):
```
I'm not crazy about having two different grids for finite build coils that need to get combined in a particular way. I think it would make more sense to just require the user to give a 3d grid if they want 3d effects (and hence if they pass in a 1d grid it just does the usual centerline only stuff)

f0uriest commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+
+    def __init__(
+        self,
+        cross_section_dims=[0.1, 0.1],
+        cross_section_shape="rectangular",
+        current=1,
+        center=[10, 0, 0],
+        normal=[0, 1, 0],
+        r_n=2,
+        modes=None,
+        basis="xyz",
+        name="",
```
minor point but i'd prefer if the ordering kept the regular FourierPlanarCoil params first and added the finite build params on at the end

ddudt replied:
Didn't we decide that all new classes should require keyword arguments going forward? So let's enforce that by adding `*` after `self`.  (And also make this change for all the other new classes added in this PR.)

f0uriest replied:
that was more for objectives, I don't think we decided for that for all classes

f0uriest commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+        )
+
+    @classmethod
+    def from_FourierPlanarCoil(
```
would also be useful to have a `to_FourierPlanarCoil` method to give you back the centerline

f0uriest replied:
we should also think about how to convert between other finite build coils, like how the regular `FourierPlanarCoil` has a `to/from_FourierXYZ` etc.

ddudt replied:
Ideas:

- circle -> rectangle can convert to a square cross-section with width=radius
- rectangle -> circle can use the average width/height as the radius

ryanwu4 replied:
I've added an overridden `to_FourierPlanarCoil`. For the latter two conversion cases (to finite build coils with other parameterizations and between cross sections), are these used commonly enough? The conversion to other finite build coils could be doable via converting to the underlying coil class and then adding finite build dimennsions again, or perhaps I can add those methods to the generic `AbstractFiniteBuildCoil` once we have more parameterizations. 

For circle <-> rectangle, I feel like the user would usually want to just reset the cross section dims with the setter method, rather than have an implicit conversion. Is that a fair assumption?

dpanici replied:
Yea I think that for converting from circle <> rectangle, I would require them to set the specific things instead of assuming defaults

f0uriest commented on desc/objectives/_coils.py:
```diff
@@ -850,6 +858,233 @@ def body(k):
+        Coil(s) that are to be optimized.
+    component :  {'mag', 'p', 'q'}, optional
+        Component of magnetic field to be targeted. Can be 'mag' for field magnitude,
+        or 'p' or 'q' for the respective dimensions of the conductor cross section.
```
do we also want an option to project onto the tangent direction?

f0uriest reviewed (commented):
This is a first pass, mostly focused on API stuff. There's a lot of good things here but I think it can be cleaned up/simplified.

One thing that felt kind of clunky was the grids for centerline vs cross section, and I wonder if we should make some new dedicated grid classes?

Right now all of our grid classes are basically assuming the coordinate system is rho,theta,zeta in an equilibrium, and we sort of hackily make that work for surfaces and curves.

A better option might be to make a dedicated `CurveGrid` class that only has a 1d set of nodes along the centerline, a `SurfaceGrid` that only has 2d nodes on a surface etc. Then we could also make a `FiniteBuildCoilGrid` or something with nodes along the centerline and cross section?

ddudt commented on desc/compute/_field.py:
I think the first few functions about curve coordinates/components can be moved to `_curve.py`, but the last few functions related to `B` can stay here.

ddudt commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+        representing the dimensions in the p and q directions, respectively.
+        Refer to fig 3 in https://arxiv.org/pdf/2310.12087 for the p and q directions.
+
+    cross_section_shape : str
```
Yeah I don't think this input is necessary, since we can assume it is circular if `cross_section_dims.size == 1` and rectangular if `cross_section_dims.size == 2` (and throw an error if size > 2)

ryanwu4 commented on desc/compute/_field.py:
```diff
@@ -3837,3 +3838,368 @@ def _L_grad_B(params, transforms, profiles, data, **kwargs):
+    params=[],
+    transforms={"grid": []},
+    profiles=[],
+    coordinates="rt",  # just two degrees of freedom
```
Is there a clearer notation I should use the denote the coordinate frame of the coil? I was mostly just trying to signify that it varies only over the first two dimensions of the grid.

ryanwu4 commented on desc/objectives/_coils.py:
```diff
@@ -850,6 +858,233 @@ def body(k):
+        from desc.coils import CoilSet, _FiniteBuildCoil
+
+        if target is None and bounds is None:
+            bounds = (-0.0, 20.0)
```
I did choose that default just based on the 20 T max field as a reasonable limit for current HTS. Just changed it to `target=0` for simplicity.

ryanwu4 commented on desc/coils.py:
```diff
@@ -1328,6 +1481,756 @@ def _check_type(coil0, coil):
+            # make params None if it's empty to correctly handle compute functions
+            if not params:
+                params = None
+        if not multifilament:
```
I'm going to implement the change you suggested where the filamentary operation is used if a 1D grid is passed and the multifilament is used if a 3D grid is passed.

ryanwu4 commented on tests/test_compute_everything.py:
```diff
@@ -171,6 +177,15 @@ def test_compute_everything():
+            current=5,
+            center=[10, 1, 3],
+            normal=[1, 2, 3],
+            r_n=[1, 2, 3],
+            modes=[0, 1, 2],
```
I just matched the `FourierPlanarCoil` implementation, I can fix all of these if its preferred but I want to be sure that I don't break things when I do that

ryanwu4 commented:
This latest commit should address all the comments above except for the analytic tangents. There's now only a 3D grid used for all the relevant compute function, but I've kept the `prep_grids ` class method just for utility. The multifilament integral and the coilset max field on coil objective are now benchmarked more rigorously as well.

dpanici commented on desc/coils.py:
```diff
@@ -1456,6 +1602,770 @@ def _check_type(coil0, coil):
+                    "The cross section grid must have endpoint=True to compute "
+                    "the self field"
+                )
+            elif xsection_grid.N > 1:
```
I am not sure if N=1 is what you want, don't you actually want `grid.num_zeta=1`?

an `N=1` LinearGrid has `2N+1` zeta planes in it

dpanici commented on desc/coils.py:
```diff
@@ -1456,6 +1602,770 @@ def _check_type(coil0, coil):
+        elif isinstance(centerline_grid, numbers.Integral):
+            centerline_grid = LinearGrid(N=centerline_grid)
+        elif isinstance(centerline_grid, LinearGrid):
+            if centerline_grid.L > 1 or centerline_grid.M > 1:
```
or L==0 and M==0

dpanici commented on desc/coils.py:
```diff
@@ -1456,6 +1602,770 @@ def _check_type(coil0, coil):
+        # pass the basis_in arg to compute_fun if necessary
+        try:
+            compute_kwargs["basis_in"] = self.basis
+        except KeyError:
```
I think this would be an AttributeError, not a KeyError

dpanici commented on desc/objectives/_coils.py:
```diff
@@ -850,6 +858,258 @@ def body(k):
+        )
+
+        self._dim_f = coilset.num_coils
+        self._constants = {
```
I think we are trying to move away from constants #1758 , so would prefer if instead these were all just attributes of the objective self (though I guess `"quad_weights"` is special as we check for it later right in`compute_scaled_error`? so maybe just leave that one but put rest as attributes of self)

dpanici commented on desc/objectives/_coils.py:
```diff
@@ -850,6 +858,258 @@ def body(k):
+                params=x,  # just this coil's parameters
+            )
+
+            quad_points = jnp.where(
```
what is this line doing?
